python/tvm/relay/backend/aot/aot.py

```diff
@@ -0,0 +1,282 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Defines the entry point into the AoT compiler.
+"""
+import ctypes
+import os
+import subprocess
+import tempfile
+import time
+
+import tvm
+from tvm import relay, get_global_func, register_func
+from tvm.relay.function import Function
+from tvm.relay.expr import Expr, Let, GlobalVar
+from tvm.relay.adt import Constructor
+from tvm.relay.expr_functor import ExprFunctor
+from tvm.relay.backend import compile_engine
+from .little_cpp import (PackedCall, CPPFunction, Invoke, Decl, CPPIf,
+                         CPPTuple, CPPMatch, CPPConstructor, CPPTupleGetItem,
+                         CPPRefCreate, CPPRefRead, CPPRefWrite)
+from . import to_source
+from .convert import convert
+
+TVM_PATH = os.environ['TVM_HOME']
+
+def must_run_process(args):
+    proc = subprocess.run(args, check=True)
+    assert proc.returncode == 0
+
+def compile_cpp(source, lib_name, flags=None, lib_path=None):
+    """
+    Compiles the given source into a C++ library
+    and returns the full path to the compiled library.
+    """
+    if flags is None:
+        flags = []
+
+    if lib_path is None:
+        lib_path = os.curdir
+
+    debug_source_path = os.path.join(lib_path, 'source.cc')
+    # Write out the file for debugging.
+    with open(debug_source_path, 'w') as source_file:
+        source_file.write(source)
+
+    # with tempfile.TmporaryDirectory() as tmpdir:
+    tmpdir = tempfile.mkdtemp(prefix="relay_aot_compiler")
+    lib_path = os.path.join(tmpdir, lib_name)
+    source_path = os.path.join(tmpdir, 'source.cc')
+    with open(source_path, 'w') as source_file:
+        source_file.write(source)
+
+    must_run_process(["clang-format", "-i", debug_source_path])
+
+    system = os.uname()[0]
+    include_paths = [
+        f"-I{TVM_PATH}/3rdparty/dmlc-core/include",
+        f"-I{TVM_PATH}/3rdparty/dlpack/include",
+        f"-I{TVM_PATH}/3rdparty/HalideIR/src",
+        f"-I{TVM_PATH}/include",
+        f"-L{TVM_PATH}/build"
+    ]
+
+    if system == 'Darwin':
```

Comment:
Yes, I think this should be written in a more maintainable manner. I will see if it can be made to programmatically match up with TVM's own C++ build configuration, for example